staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

```diff
@@ -6893,16 +7021,365 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
+								{
+									Rule:    "size(self) > 0",
+									Message: "scoped field should contain more than 0 element.",
+								},
+							},
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			opts: validationOptions{
+				requireStructuralSchema: true,
+			},
+		},
+		{
+			name: "valid x-kubernetes-validations for array of object",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Items: &apiextensions.JSONSchemaPropsOrArray{
+						Schema: &apiextensions.JSONSchemaProps{
+							Type: "array",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule:    "self[0].nestedObj.val > 0",
+									Message: "val should be greater than 0.",
+								},
+							},
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "object",
+									Properties: map[string]apiextensions.JSONSchemaProps{
+										"nestedObj": {
+											Type: "object",
+											Properties: map[string]apiextensions.JSONSchemaProps{
+												"val": {
+													Type: "integer",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			opts: validationOptions{
+				requireStructuralSchema: true,
+			},
+		},
+		{
+			name: "valid x-kubernetes-validations for escaping",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					XValidations: apiextensions.ValidationRules{
+						{
+							Rule: "self.__if__ > 0",
+						},
+						{
+							Rule: "self.__namespace__ > 0",
+						},
+						{
+							Rule: "self.self > 0",
+						},
+						{
+							Rule: "self.int > 0",
+						},
+					},
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"if": {
+							Type: "integer",
+						},
+						"namespace": {
+							Type: "integer",
+						},
+						"self": {
+							Type: "integer",
+						},
+						"int": {
+							Type: "integer",
+						},
+					},
+				},
+			},
+			opts: validationOptions{
+				requireStructuralSchema: true,
+			},
+		},
+		{
+			name: "invalid x-kubernetes-validations for escaping",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					XValidations: apiextensions.ValidationRules{
+						{
+							Rule: "if > 0",
+						},
+						{
+							Rule: "namespace > 0",
+						},
+						{
+							Rule: "unknownPro > 0",
+						},
+						{
+							Rule: "self > 0",
+						},
+						{
+							Rule: "int > 0",
+						},
+					},
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"if": {
+							Type: "integer",
+						},
+						"namespace": {
+							Type: "integer",
+						},
+						"self": {
```

Comment:
am curious: does `size(self)` work? I have seen ppl modelling unions using such an expression, i.e. `size(volume) == 1` because only `empty`, `hostPath` or one of the other could be set at one time.

Reply:
Size is not allowed on objects with properties (the CEL spec calls these "messages"). It's only allowed on strings, bytes, lists and maps:

```
(string) -> int | string length
(bytes) -> int | bytes length
(list(A)) -> int | list size. Time cost proportional to the length of the list.
(map(A, B)) -> int | map size. Time cost proportional to the number of entries.
```

xref https://github.com/google/cel-spec/blob/master/doc/langdef.md#list-of-standard-definitions

Reply:
good reference, thanks. it looks like the openapi `maxProperties` validation attribute could be used for the union-of validation